extension/jaegerremotesampling/internal/remote_strategy_cache.go

```diff
@@ -0,0 +1,134 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/jaegerremotesampling/internal"
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/jaegertracing/jaeger/thrift-gen/sampling"
+	"github.com/tilinna/clock"
+)
+
+type serviceStrategyCache interface {
+	get(ctx context.Context, serviceName string) (*sampling.SamplingStrategyResponse, bool)
+	put(ctx context.Context, serviceName string, response *sampling.SamplingStrategyResponse)
+	Close() error
+}
+
+// serviceStrategyCacheEntry is a timestamped sampling strategy response
+type serviceStrategyCacheEntry struct {
+	retrievedAt      time.Time
+	strategyResponse *sampling.SamplingStrategyResponse
+}
+
+// serviceStrategyTTLCache is a naive in-memory TTL serviceStrategyTTLCache of service-specific sampling strategies
+// returned from the remote source. Each cached item has its own TTL used to determine whether it is valid for read
+// usage (based on the time of write).
+type serviceStrategyTTLCache struct {
+	itemTTL time.Duration
+
+	stopCh chan struct{}
+	rw     sync.RWMutex
+	items  map[string]serviceStrategyCacheEntry
+}
+
+// Initial size of cache's underlying map
+const initialRemoteResponseCacheSize = 32
+
+// Minimum time between runs of "clean up" of all cache-allocated data.
+// Used if the itemTTL is lower than this.
+const minCleanupSchedulingPeriod = 5 * time.Minute
+
+func newServiceStrategyCache(itemTTL time.Duration) serviceStrategyCache {
+	result := &serviceStrategyTTLCache{
+		itemTTL: itemTTL,
+		items:   make(map[string]serviceStrategyCacheEntry, initialRemoteResponseCacheSize),
+		stopCh:  make(chan struct{}),
+	}
+
+	// Launches a "cleaner" goroutine that naively blows away cached data
+	cleanerSchedulingPeriod := itemTTL
+	if itemTTL < minCleanupSchedulingPeriod {
+		cleanerSchedulingPeriod = minCleanupSchedulingPeriod
+	}
+	go result.periodicallyClearCache(context.Background(), cleanerSchedulingPeriod)
+	return result
+}
+
+// get returns a cached sampling strategy if one is present and is no older than the serviceStrategyTTLCache's per-item TTL.
+func (c *serviceStrategyTTLCache) get(
+	ctx context.Context,
+	serviceName string,
+) (*sampling.SamplingStrategyResponse, bool) {
+	c.rw.RLock()
+	defer c.rw.RUnlock()
```

Comment:
do you need to defer? Can you let go of the lock a bit earlier here?

Reply:
don't need to, it's just a pattern I've seen (Go newbie here). In this case the subsequent branches are almost all immediate returns based on some in-memory state so I don't think lock contention is that big of a deal, so I lean toward the code cleanliness / lazy approach of treating the entire function as synchronized rather than having some fine grained locking around critical sections inside of it.